apache/incubator-mxnet
# [WIP][MXNET-107] Fused LSTM implementation for CPU

## Description
In this PR, a fused LSTM operator for CPU is implemented. More supports for other RNN variants are WIP and will be submitted in other PRs.

## Feature changes
### New features
* Fused LSTM implemention, including both forward and backward computation.
* Share the same frontend interfaces with current `sym.RNN` operator
* Share the same algorithm and input layout with current `sym.RNN` operator
* ~~Refactor code and register it with NNVM interfaces~~
* Support both FP32 and FP64 inputs, more types will be supported later, such as int8
* Provide more extensible APIs for other RNN variants(vanilla RNN/GRU)


## Unit-test changes
* Create new test `test_lstm` and `test_lstm_bidirectional`in `tests/python/unittests/test_operator.py` 
* Check consistency with original LSTMCell implementation

## Performance
We have tested performance of `sym.RNN` and `rnn.LSTMCell` on our local Skylake-8180 with 2 Sockets and 56 cores. Use MKL as blas lib in this performance test.
Test input size is from DS2 default parameters(seq_length = 300, batch_size = 20, input_size = 800, hidden_size = 800).

**single layer measurement**:

| API                  | Inference time(fwd, sec)  |  Training time(fwd + bwd, sec)           |
| --------           | :-----:                              | :----:                                   |
| rnn.LSTMCell  | 0.106902                       |  0.273108                            |
| #9977             | 0.050126                      |    ---                                   |
| this PR            | 0.050668                      |  0.130266                               |
| speedup         | 2.1x                              |  2.1x                                    |

**multi layer measurement**: num_layers=5

| API                  | Inference time(fwd, sec)  |  Training time(fwd + bwd, sec)           |
|  :--------:         | :-----:                              | :----:                                   |
| rnn.LSTMCell | 0.532034                       | 1.546486                          |
| sym.RNN(#9977) | 0.18641                         |    ---                                   |
| sym.RNN(this PR) | 0.190032                     |   0.619439                               |
| rnn.LSTMCell(cuda) | 0.231355                     | 0.785780                         |
| sym.RNN(cudnn)  | 0.060647                     |   0.161115                               |
| speedup  <br> #10104 /LSTMCell        | 285.41%                             |  249.66%                                  |
| speedup  <br> #10104 /LSTMCell(cuda) | 124.09%                          |  126.85%                                  |
| speedup <br>  #10104 / sym.RNN(cudnn) | 32.53%                        |  26.01%                                    |


## Opens
* [ ] ~~Fix cudnn registeration in this PR~~
* [x] Add multi-layer and bidirectional support for LSTM.
* [x] Support gluon interfaces
* [ ] fix NNVM registration
* [ ] Other RNN variants (will be added in in other PRs)
* [ ] Add dropout support(in other PRs)


## Checklist ##
### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
